File: algorithms-practice/dynamic_programming/lis.cpp
// Longest Increasing Subsequence (Patience Sorting)
// Time: O(n log n)
// Space: O(n)
#include <bits/stdc++.h>
using namespace std;

int lengthOfLIS(const vector<int>& nums){
    vector<int> tails;
    for (int x : nums){
        auto it = lower_bound(tails.begin(), tails.end(), x);
        if (it == tails.end()) tails.push_back(x);
        else *it = x;
    }
    return (int)tails.size();
}

int main(){
    vector<int> nums = {10,9,2,5,3,7,101,18};
    cout << lengthOfLIS(nums) << "\n"; // 4
    return 0;
}

File: algorithms-practice/dynamic_programming/lis.py
"""
Longest Increasing Subsequence (Patience Sorting)
- Time: O(n log n)
- Space: O(n)
"""
from bisect import bisect_left
from typing import List


def length_of_lis(nums: List[int]) -> int:
    tails: List[int] = []
    for x in nums:
        i = bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)


if __name__ == "__main__":
    print(length_of_lis([10,9,2,5,3,7,101,18]))  # 4
